### ISSUE TITLE
Physics/AI error after staying too long on a map

### ISSUE THREAD

**Issue by Ch0wW** [skullernet/q2pro#185]
It looks like after staying more than a week on a same map (maybe it's more), I noticed that many lift physics were out of place (=going far higher than the map intended), aswell as seeing super super slow enemies.

Using Q2Pro's latest revision, both for server (Debian 9.X) & client (Windows 10). 
Server is using using YamagiQ2's dll binaries for baseq2. 

![quake000](https://user-images.githubusercontent.com/2938006/55277816-881bab00-5304-11e9-9af6-46db258aaf7e.jpg)
![quake001](https://user-images.githubusercontent.com/2938006/55277817-881bab00-5304-11e9-827c-6ad791a5c1b8.jpg)
![quake002](https://user-images.githubusercontent.com/2938006/55277818-881bab00-5304-11e9-9904-ea23895a9bbc.jpg)
![quake003](https://user-images.githubusercontent.com/2938006/55277819-88b44180-5304-11e9-96ff-248e4d26ab45.jpg)

**Comment by tojestzart** [comment]
@Ch0wW  cant understand those pics from single mode but last one showing definitley a bug, same was on q2dm1 ages ago, depending from [MOD]  and yes its atleast moving way more as expected , before it would rocket to the sky ask @Yamagi whats wrong

**Comment by Ch0wW** [comment]
What I mean by "slow" enemies is, their behaviour isn't that reactive at all compared to what you'd see in a regular game play. Their movement are sluggish, and can be worsened if you leave the server even longer on the same map.

I tried to play my server on Q2Pro latest version, and the switch port of YamagiQ2, and both are having the same issue on my server.

On that Ammo Depot lava bridge part, you'd notice there's a gap seen on the lava bridge, that shouldn't be present at all.

Also, I noticed that reloading the map from the server fixes it... Even if all monsters reappear for some reason (common vanilla q2 bug ?)

**Comment by Yamagi** [comment]
I'm somewhat astonished that it takes one week for things do go haywire. There's some counters in the game code, that may wrap after some time. Some calculation are done in floating math, so errors may accumulate until they become too big... Loading a new map resets the game, cleaning all those errors out.

Quake II is just not designed to run very long times, the original client and Win98 were unstable enough to force restarts every now and then. Nevertheless, if someone is able to figure out what exactly goes wrong we (yquake2) would be happy to try to solve it.

**Comment by skullernet** [comment]
This is a known Q2 bug, stemming from the fact level.time and associated timestamp fields in the code are floating point values. As magnitude of level.time increases, calculations become less precise, and constructions like this that need to wait exactly N frames (with N=1 typically) start to break approximately after one week:
```
ent->nextthink = level.time + n * FRAMETIME;
```
Solution is to convert nextthink and other timers to use level.framenum, which is an integer. This would allow the server to stay on the same map for more than 6 years until it overflows (assuming standard 10 FPS server frame rate).

This bug has already been fixed in OpenTDM and OpenFFA, but I've never bothered to fix baseq2 as single player servers rarely stay on the same level for too long. I've added some commits to address this issue to master (hopefully not breaking anything else given how invasive the changes are).
